implementation/rust-embedded-library.tex

```diff
@@ -1,65 +1,62 @@
+\gls{rcl} defines functionality to work with atomic values and cells along with exposing compiler \concept{intrinsics} which are used for example to provide \concept{volatile} access to memory.
+The \gls{rcl} does not have any library dependencies, but in order to use the library without \gls{rsl} a few definitions given in \autoref{tab:core:definitions} must exists.
 
 \begin{table}[H]
-  \begin{tabular}{l}
-    Binary Heap \\
-    Binary Tree Map \\
-    Binary Tree Set \\
-    Bit Set \\
-    Bit Vector \\
-    Enum Set \\
-    Linked List \\
-    Vector \\
-    Vector Dequeue \\
-    Vector Map \\
-    String \\
+  \begin{tabular}{l l}
+    \code{memcpy, memcmp, memset} & basic memory management \\
+    \code{rust\_begin\_unwind} & handling panicing \\
   \end{tabular}
-  \caption{\rust Collection Library}
-  \label{tab:rust:collections}
+  \caption{}
+  \label{tab:core:definitions}
 \end{table}
 
-We look at some of the data structures in this section.
-
-\subsubsection{Vector}
+The memory management functions given in \autoref{tab:core:definitions} are provided by \lib{newlib} and are exposed thought the \lib{startup} library described in \autoref{sec:startup}.
+The \code{rust\_begin\_unwind} is also defined in \lib{startup}, but the implementation is only an infinite loop to aid debugging.
+In contrast the definition of \code{rust\_begin\_unwind} given in {\std} will abort the program and print an error message.
 
-The Vector is an important data structure both in direct usage but also as a building block for other data structures.
-In essence it is a growable heap allocated list of values of the same type.
-
-\subsubsection{String}
-
-In \autoref{par:rust:str} we discussed the \concept{string slice}.
-The \code{String} defined in the collection library is a heap allocated mutable string.
-
-\subsection{libc}
+\subsection{The Allocation Library}
+\label{sec:rust:allocation}
 
-The heap allocation library described in \autoref{sec:rust:allocation} depends on the \lib{libc} crate abstraction.
-The \lib{libc} library provided in the standard library provides bindings for the {\C} standard library and platform specific libraries.
-\gls{rel} depends on \lib{libc} for the low level memory operations \func{malloc}, \func{realloc} and \func{free}.
-The details of this implementation is found in \autoref{sec:impl:libc}. \todo{Update reference to implementation chapter part about libc implementation}
+Heap allocation is introduced in a library called \lib{alloc}.
+The library defines the owned pointer \code{Box} which is {\rust}'s main means of allocating memory on the heap.
+In addition the allocation library defines the types \code{Rc} and \code{Arc} which are {\rust}'s reference counted and atomically reference counted heap pointers.
+The allocation library is by default dependent on \lib{libc}, but by suppling the \flag{--cfg feature="external\_funcs"} this dependency can be broken.
```

Comment:
Kanskje endre til: `on libc, but this dependency can be broken by supplying the --cfg ... flag to the compilation procedure.`